### ISSUE TITLE
Provide mechanism for creating new boxed version of a COM class casted as a particular interface. 

### ISSUE THREAD

**Issue by cmyr** [microsoft/com-rs#161]
I'm playing around with the latest revision of this crate (currently 3693ab2e1); I'm trying to implement some interfaces for a custom type, and I'm trying to figure out how to get one of the interface I implement once I have created an instance my concrete 'class'.

afaict, the only way I can do this is by using `query_interface`, and doing a bunch of pointer casts. I see that there's a `get_interface` method defined on the interface type, but it isn't defined on my class, so it doesn't seem like I can use it? The best I can do so far in terms of some kind of general solution is to use a macro like,

```rust
macro_rules! get_interface {
    ($item:expr, $interface:ty) => {{
        let mut ppv: Option<$interface> = None;
        let iid = &<$interface as com::Interface>::IID as *const _;
        let hr = unsafe {
            $item.query_interface(iid, &mut ppv as *mut _ as *mut *mut c_void)
        };
        if SUCCEEDED(hr) {
            ppv
        } else {
            None
        }
    }}
}
```

and then I can call this like, `let interface = get_interface(my_type, IDWriteFontFile)`, but it doesn't feel like a great solution, but since query_interface isn't part of a trait I don't see an alternative?

I'm probably missing something obvious, but any pointers here would be appreciated. :)

**Comment by rylev** [comment]
Hey there!👋 Just to clarify, this is inside a method implementation for a COM class and you want to convert `self` (the this pointer) to an interface pointer? Can I ask why? You have access to all the methods on `self` already. Are you trying to return a COM interface pointer from a method?

**Comment by MarijnS95** [comment]
@rylev I mentioned this issue [before](https://github.com/microsoft/com-rs/pull/156#discussion_r460243317) in the relevant PR and you made an [issue for it](https://github.com/microsoft/com-rs/issues/160) but was quickly closed.

My usecase is implementing an interface in Rust through a class and passing that instance down to a COM function accepting a pointer to that interface. Currently it doesn't seem possible to (implicitly) convert that class to an interface. Not to mention the lifetime issues it brings.

For reference, we're defining a class implementation [here](https://github.com/Traverse-Research/hassle-rs/blob/4849fefa44944b55d7db93bbbb28955670d36ab2/src/wrapper.rs#L120) and passing it through a COM function as interface pointer [here](https://github.com/Traverse-Research/hassle-rs/blob/4849fefa44944b55d7db93bbbb28955670d36ab2/src/wrapper.rs#L308).

**Comment by cmyr** [comment]
Hi @rylev! Here's a concrete example (and with the disclaimer that I have approximately zero experience with COM, previously): I'm trying to implement a bunch of dwrite interfaces, which (as you mention) require me to return a com type via pointer.

Here's some simple example code; in `get_current_font_file` I need to create and return an `IDWriteFontFile`; I implement this interface on my `PietFontFile` type.

```rust
com::interfaces! {
    #[uuid("EFF8970E-C50F-45E0-9284-291CE5A6F771")]
    pub unsafe interface IDWriteFontFileEnumerator: IUnknown {
        fn move_next(&self, has_current: *mut BOOL) -> HRESULT;
        fn get_current_font_file(&self, file: *mut Option<IDWriteFontFile>) -> HRESULT;
    }

    #[uuid("739d886a-cef5-47dc-8769-1a8b41bebbb0")]
    pub unsafe interface IDWriteFontFile: IUnknown {
        fn get_reference_key(
            &self,
            key: *mut *const c_void,
            key_size: *mut UINT32,
        ) -> HRESULT;
        fn get_loader(&self, loader: *mut Option<IDWriteFontFileLoader>) -> HRESULT;
        fn analyze(
            &self,
            is_supported_file_type: *mut BOOL,
            file_type: *mut DWRITE_FONT_FILE_TYPE,
            face_type: *mut DWRITE_FONT_FACE_TYPE,
            number_of_faces: *mut UINT32,
        ) -> HRESULT;
    }
}

// details omitted for brevity
com::class! {
    pub class PietFontFile: IDWriteFontFile {}
    impl IDWriteFontFile for PietFontFile { }
}

com::class! {
    pub class PietFontFileEnumerator: IDWriteFontFileEnumerator {
        files: Rc<Vec<FontData>>,
        idx: Cell<Option<usize>>,
    }

    impl IDWriteFontFileEnumerator for PietFontFileEnumerator {
        fn move_next(&self, has_current: *mut BOOL) -> HRESULT {
            let next_idx = self.idx.get().map(|n| n + 1).unwrap_or(0);
            let has_item = if next_idx < self.files.len() {
                TRUE
            } else {
                FALSE
            };
            unsafe { *has_current = has_item; }
            self.idx.set(Some(next_idx));
            S_OK
        }

        fn get_current_font_file(&self, file: *mut Option<IDWriteFontFile>) -> HRESULT {
                let font_file = PietFontFile::default();
                unsafe { *file = get_interface!(font_file, IDWriteFontFile); }
            S_OK
        }
    }
}


```

**Comment by rylev** [comment]
Thanks so much for this! What you're trying to do, we don't yet support, but we absolutely should. As the code is currently written, you'll segfault as you stack allocate PietFontFile which gets dropped at the end of `get_current_font_file` , but you're handing out pointers to it through the `file` param. You need to heap allocate first. We should provide a convenience for heap allocating the class and querying for an interface. We currently do this for you in the `DllGetClassObject` helper but as you shown here, it's useful to be able to do this in methods. I also think we need to prevent calling the method from a non-boxed self.

**Comment by rylev** [comment]
Here's a small playground showing how to prevent stack allocated classes from calling methods. We need to implement this for all COM classes: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d8fc4365bfcd10452c68f35bb197f003

**Comment by cmyr** [comment]
I assume I can work around this for the time being by boxing and `mem::forgetting` my class?

**Comment by rylev** [comment]
Yea you basically want to recreate this: https://github.com/microsoft/com-rs/blob/master/src/production/registration.rs#L209-L216

I'm going to change the title of this issue to reflect what needs to be done.

**Comment by cmyr** [comment]
what's up with the retain/release around query_interface?

**Comment by MarijnS95** [comment]
> I'm going to change the title of this issue to reflect what needs to be done.

@rylev would you mind reopening https://github.com/microsoft/com-rs/issues/160 though? That issue still stands as the original comment from @cmyr requires manual wrapping of `query_interface`.

Re-asking https://github.com/microsoft/com-rs/pull/156#discussion_r465243340, why is https://github.com/microsoft/com-rs/commit/27daf863e7dd5d65c5dd87ea0a78828f7458bf81 necessary?

> what's up with the retain/release around query_interface?

There seem to be some lifetime issues as mentioned [here](https://github.com/microsoft/com-rs/pull/156#discussion_r460243317). @rylev would it be make sense to attach lifetimes (using `PhantomData`) to interface and class structs? `get_interface` on interfaces currently drops that as well, seemingly relying on just refcounting. This may be intended, especially if an externally managed interface is used (which relies on refcounting already, as per the COM spec).

> Though when implementing this locally (copypasting `IUnknown::get_interface`) it's running into lifetime issues because the returned interface deals with refcnt while a generated class does not fully:
> 
> * Should the refcnt start with 1?
> * Should every class have `impl Drop for #name` and call `.release()`, just like interface implementations?

**Comment by kennykerr** [comment]
Ref counts should always be initialized with 1 so that the creator/caller owns the originating ref count and can either transfer ownership of the ref count or release it. Not doing so has a long history of introducing ref counting bugs in C++. Either way, calls using that pointer (such as `query_interface`) do not need to be stabilized with a wrapping `add_ref`/`release`.

**Comment by MarijnS95** [comment]
> Ref counts should always be initialized with 1 so that the creator/caller owns the originating ref count and can either transfer ownership of the ref count or release it. Not doing so has a long history of introducing ref counting bugs in C++

This is a bug then :slightly_smiling_face::

https://github.com/microsoft/com-rs/blob/3693ab2e198709f3eb4c705facc3bbdd1fb5ca69/macros/support/src/class/class_constructor.rs#L42-L47

Is that also why wrapping occurs here?

https://github.com/microsoft/com-rs/blob/3693ab2e198709f3eb4c705facc3bbdd1fb5ca69/macros/support/src/class/class_factory.rs#L27-L29

> Either way, calls using that pointer (such as `query_interface`) do not need to be stabilized with a wrapping `add_ref`/`release`.

Yup, because `QueryInterface` does that [for you](https://docs.microsoft.com/en-us/cpp/atl/queryinterface?view=vs-2019).

**Comment by rylev** [comment]
I hope #163 solves all of these issues. 

@MarijnS95, we don't need to call `add_ref` in the `From` and `Deref` impls because those don't produce new handles to an interface. 

The `From` impl consumes the old interface handle and produces a new one (not running the destructor for the old one). This is effectively just treating the old interface as the new interface. There's no new handle so the ref count doesn't need to increase. 

The `Deref` impl on the other hand takes a reference to an interface and produces a reference to another interface whose lifetime is tied to the first. Therefore it's impossible for the second interface to exist after the first is dropped, so there's no need to explicitly keep tabs of it through reference counting because the Rust lifetime system does this for us.

**Comment by MarijnS95** [comment]
> @MarijnS95, we don't need to call `add_ref` in the `From` and `Deref` impls because those don't produce new handles to an interface.

The two explanations make sense, thanks for diving into that.

> I don't believe there's a need for a lifetime param on the safe QI as it will just produce a new interface handle which increases the reference count.

I see now; COM objects have a different lifetime scope outside of Rust, using boxes/pin and an internal reference count only makes sense.

That said and attempt to use the new `query` function from #163 - wrapping the class in a Box and Pin manually - ends with a segfault because the object and/or vtable gets broken (freed?) somewhere before going into native code. I'm not far enough into unsafe Rust to fully understand how and where these objects are dropped, and how `ManuallyDrop` plays with that. On top of that it seems like com-rs is getting increasingly complex and hard to follow. Turns out  class instances have to be threaded through `fn allocate`; the wrapping in `ManuallyDrop` here likely performing the magic trick.
Is this a case of me using com-rs/Rust wrong, or com-rs not fully protecting against all implicit drop cases yet?

Which is why I want to repeat my request to add an example that passes a `class` as `interface` through a function, so that it's clear how the API is intended to be used. In addition that'll help creating simple reproduction cases for borrow/refcount/drop issues.

**Comment by rylev** [comment]
I left a comment in the PR explicitly about the case you're running into though I thought it wouldn't lead to seg-fault (just leaking memory). I'm still trying to figure out the best way to handle that. I will definitely be adding an example. Thanks for living on the edge and testing all of this out. COM itself is not really possible to make completely safe, so I'm doing my best to make it as safe as possible.

**Comment by MarijnS95** [comment]
> I left a comment in the PR explicitly about the case you're running into though I thought it wouldn't lead to seg-fault (just leaking memory).

After putting a log line in place that specific piece doesn't even seem to get reached, nor do the other drops seem to get hit. The segfault happens in the native library we're using because the vtable the object points to (if it's even the original vtable) is garbled.

I've pushed the repro with manual Box+Pin [here](https://github.com/Traverse-Research/hassle-rs/commit/fa7653d808fe2b6b10918e24b76cb27a40772050) assuming you have access to a dxcompiler library. Note that the code doesn't compile on Windows yet because `#[cfg]` isn't supported within `class!`, but you can easily comment out the conflicting function signatures.

> Thanks for living on the edge and testing all of this out. COM itself is not really possible to make completely safe, so I'm doing my best to make it as safe as possible.

Doing my best to replace deprecated `com` as soon as the API here stabilizes and suffices our needs, including Linux support. Which, by the way, suffers small breakage because the example server+client are now tested in CI but incompatible with "inproc-only" support for Linux (ie. the interfaces and passing COM objects around without Win32 functionality). I have yet to disable those and update the PR.

Would you still mind to explain the necessity of https://github.com/microsoft/com-rs/commit/27daf863e7dd5d65c5dd87ea0a78828f7458bf81? It seems `Deref` to the `Super` (`IUnknown`) interface already implicitly takes care of this?

**Comment by rylev** [comment]
> The segfault happens in the native library we're using because the vtable the object points to (if it's even the original vtable) is garbled.

Can you explain a bit more what you mean by this?

> Doing my best to replace deprecated com as soon as the API here stabilizes and suffices our needs, including Linux support.

Once we get this in we can turn on experimental Linux support

> Would you still mind to explain the necessity of 27daf86? It seems Deref to the Super (IUnknown) interface already implicitly takes care of this?

Yes, an interface can query for other other interfaces that are not directly in its parental hierarchy. If class `GoldenRetriever` implments `IDog` which has a parent interface `IAnimal`, and it also implements `IFriend`, then if I have a handle to `IDog`, I can deref directly to `IAnimal`, but I need to use `query_interface` (or the safe `query`) if I want to go from `IDog` to `IFriend` since `IFriend` is a totally separate interface and not in the same hierarchy as `IDog`. Similarly if I want to downcast from `IAnimal` to `IDog` I need to use `query`. This is because not all classes that implement `IAnimal` will implement `IDog` (some might implement `ICat` for instance).

**Comment by MarijnS95** [comment]
> Can you explain a bit more what you mean by this?

According to GDB the VTable contains random values, and dxcompiler fails when attempting to call `->AddRef` which is a garbage pointer. If you want I can paste some GDB snippets showing what the data looks like.


> > Would you still mind to explain [...]
> 
> Yes [...]

I didn't mean `Deref` to other interfaces but to `IUnknown` specifically. Prior to that commit I was already able to call `.get_interface` on any interface because of the `Deref` to `IUnknown`.

**Comment by rylev** [comment]
You're totally right about there not being a need for that extra `safe_query` function. I've removed it. 

I'm not sure what the issue with the vtable is. It should now not be possible (or rather very hard) to stack allocate the class so calling `Box::pin` manually is not really a possibility anymore. You should use the `MyClass::allocate` function for instantiating classes. I'll work on creating docs for this.

**Comment by MarijnS95** [comment]
> I'm not sure what the issue with the vtable is. It should now not be possible (or rather very hard) to stack allocate the class so calling `Box::pin` manually is not really a possibility anymore. You should use the `MyClass::allocate` function for instantiating classes. I'll work on creating docs for this.

Using `allocate` indeed solved the issue, but the main point seems that users might still like to move their objects around or wrap them in specific ways inadvertently resulting in an implicit drop not caught by com-rs causing undefined behaviour.

**Comment by rylev** [comment]
@MarijnS95 the various `alloc` functions return a Pin<Box<T>> so you can't move out from it. It is _technically_ possible to construct the class object on the stack though it is quite difficult. Can you think of an example of how to get around the restrictions?

**Comment by MarijnS95** [comment]
> the various `alloc` functions return a Pin<Box>

Then this becomes less relevant now that the `allocate` function was changed to return that instead of a stack-allocated object. We don't do any wrapping at the moment but I'll let you know if we come across something that is wrongly dropped (twice).
